include/alpaka/atomic/AtomicOaccBuiltIn.hpp

```diff
@@ -0,0 +1,388 @@
+// #if _OPENACC >= 201510
+
+        //#############################################################################
+        //! The OpenACC accelerators atomic operation: ADD
+        template<
+            typename T,
+            typename THierarchy>
+        struct AtomicOp<
+            AtomicAdd,
+            AtomicOaccBuiltIn,
+            T,
+            THierarchy>
+        {
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST_ACC static auto atomicOp(
+                AtomicOaccBuiltIn const &,
+                T * const addr,
+                T const & value)
+            -> T
+            {
+                T old;
+                auto & ref(*addr);
+                // atomically update ref, but capture the original value in old
+                #pragma acc atomic capture
+                {
+                    old = ref;
+                    ref += value;
+                }
+                return old;
+            }
+        };
+
+        //#############################################################################
+        //! The OpenACC accelerators atomic operation: SUB
+        template<
+            typename T,
+            typename THierarchy>
+        struct AtomicOp<
+            AtomicSub,
+            AtomicOaccBuiltIn,
+            T,
+            THierarchy>
+        {
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST_ACC static auto atomicOp(
+                AtomicOaccBuiltIn const &,
+                T * const addr,
+                T const & value)
+            -> T
+            {
+                T old;
+                auto & ref(*addr);
+                // atomically update ref, but capture the original value in old
+                #pragma acc atomic capture
+                {
+                    old = ref;
+                    ref -= value;
+                }
+                return old;
+            }
+        };
+
+        //#############################################################################
+        //! The OpenACC accelerators atomic operation: EXCH
+        template<
+            typename T,
+            typename THierarchy>
+        struct AtomicOp<
+            AtomicExch,
+            AtomicOaccBuiltIn,
+            T,
+            THierarchy>
+        {
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST_ACC static auto atomicOp(
+                AtomicOaccBuiltIn const &,
+                T * const addr,
+                T const & value)
+            -> T
+            {
+                T old;
+                auto & ref(*addr);
+                // atomically update ref, but capture the original value in old
+#if !BOOST_COMP_PGI || defined TPR28628 // triggers PGI TPR28628, not atomic until fixed
+                #pragma acc atomic capture
+#endif
```

Comment:
Do I understand correctly, that there is a compiler with a specific version where `alpaka::atomicExch` will silently not be atomic?
I think this should be at least a warning or even a hard error.

Reply:
We need this to compile work in an experimental PIConGPU version. This whole backend is experimental because there is no compiler support. The affected compiler version cannot be considered production-ready anyway. Once there is a working compiler, the old compiler version can just be declared incompatible altogether.

Reply:
Well, if this PR is merged, this version becomes part of a public alpaka release and people are going to download and try it. I would be discouraged as a client if that caused a bug in my code, even if it's in 2 years from.

Could you please add something like a `#pragma message("Atomic exchange will not be atomic. Sorry :/")`?

Reply:
Ok. Why not #warning ?

Reply:
Because #warning is not supported by MSVC and will give you a compile error. Which would also be fine in this case :D #pargma message should work on MSVC, GCC and clang, and if not supported, must be ignored.

Reply:
In this case, we haven an issue here: https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/example/ExampleDefaultAcc.hpp#L48

Will this message be a warning and be elevated to an error by `-Werror`?

Reply:
This is a hard error on MSVC and I was too lazy yet to open an issue.